### PR TITLE
Fix an error in cluster fault rescheduling

### PR DESCRIPTION
**Pull request by XiShanYongYe-Chang**
Signed-off-by: changzhen <changzhen5@huawei.com>

**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:

Set flag `--failover=true` in karmada-scheduler.

User create propagationPolicy to propagate resource:

```yaml
apiVersion: policy.karmada.io/v1alpha1
kind: PropagationPolicy
metadata:
  name: example-policy
spec:
  ...
  placement:
    clusterAffinity:
      clusterNames:
        - member1
        - member2
```
Rescheduling is triggered when `member2` cluster becomes not ready.

CandidateClusters will be calculate in:
https://github.com/karmada-io/karmada/blob/77c99f18b0a3ed58c4609566e4ef8588296e5fab/pkg/scheduler/scheduler.go#L734

The `candidateClusters` isn't the real candidate clusters which conform with propagationPolicy. This error causes the propagationPolicy to be incorrectly updated to:

```yaml
apiVersion: policy.karmada.io/v1alpha1
kind: PropagationPolicy
metadata:
  name: example-policy
spec:
  ...
  placement:
    clusterAffinity:
      clusterNames:
        - member1
```

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/util/binding.go [hunk 1]
```diff
@@ -45,3 +47,13 @@ func GetSumOfReplicas(clusters []workv1alpha1.TargetCluster) int32 {
 	}
 	return replicasSum
 }
+
+// GetClusterSetString will get the sets.String from binding's clusters
```
Comment (RainbowMango) [inline pkg/util/binding.go:51]:
```suggestion
// GetClusterSetString will get the sets.String from a cluster slice.
```

File: pkg/scheduler/scheduler.go [hunk 2]
```diff
@@ -770,13 +723,77 @@ func (s *Scheduler) rescheduleResourceBinding(resourceBinding *workv1alpha1.Reso
...
+}
+
+func getAvailableCluster(clusters []workv1alpha1.TargetCluster, readyClusterInfos []*framework.ClusterInfo) sets.String {
+	boundClusters := util.GetClusterSetString(clusters)
+	readyClusters := getReadyCluster(readyClusterInfos)
+
+	return readyClusters.Difference(boundClusters)
+}
+
+func getReadyCluster(readyClusterInfos []*framework.ClusterInfo) sets.String {
```
Comment (RainbowMango) [inline pkg/scheduler/scheduler.go:747]: The function name is misleading.

File: pkg/scheduler/scheduler.go [hunk 3]
```diff
@@ -770,13 +723,77 @@ func (s *Scheduler) rescheduleResourceBinding(resourceBinding *workv1alpha1.Reso
...
-	_, err = s.KarmadaClient.WorkV1alpha1().ResourceBindings(ns).Update(context.TODO(), resourceBinding, metav1.UpdateOptions{})
+	_, err = s.KarmadaClient.WorkV1alpha1().ResourceBindings(resourceBinding.Namespace).Update(context.TODO(), resourceBinding, metav1.UpdateOptions{})
 	if err != nil {
 		return err
 	}
 	return nil
 }
 
+func getReservedCluster(clusters []workv1alpha1.TargetCluster, readyClusterInfos []*framework.ClusterInfo) sets.String {
+	boundClusters := util.GetClusterSetString(clusters)
```
Comment (RainbowMango) [inline pkg/scheduler/scheduler.go:734]:
```suggestion
	total := util.GetClusterSetString(clusters)
```
?

File: pkg/util/binding.go [hunk 4]
```diff
@@ -45,3 +47,13 @@ func GetSumOfReplicas(clusters []workv1alpha1.TargetCluster) int32 {
...
+func GetClusterSetString(clusters []workv1alpha1.TargetCluster) sets.String {
```
Comment (RainbowMango) [inline pkg/util/binding.go:52]: Can we use something like `covert` or `transxxx` name? `get` is too common.

File: pkg/util/binding.go (hunk 1 above)
Comment (XiShanYongYe-Chang) [reply pkg/util/binding.go]: I modify func name to `ConvertToClusterNames`, and modify comment to:
```
// ConvertToClusterNames will convert a cluster slice to clusterName's sets.String
```

File: pkg/scheduler/scheduler.go (hunk 2 above)
Comment (XiShanYongYe-Chang) [reply pkg/scheduler/scheduler.go]: I move this func to `/pkg/scheduler/cache/snapshot.go`, and rename it to `GetReadyClusterNames`.

File: pkg/scheduler/scheduler.go (hunk 3 above)
Comment (XiShanYongYe-Chang) [reply pkg/scheduler/scheduler.go]: Done.

File: pkg/scheduler/scheduler.go [hunk 5]
```diff
@@ -770,13 +723,71 @@ func (s *Scheduler) rescheduleResourceBinding(resourceBinding *workv1alpha1.Reso
...
 	return nil
 }
 
+func calcReservedCluster(total, ready sets.String) sets.String {
+	return total.Difference(total.Difference(ready))
+}
+
+func calcAvailableCluster(total, ready sets.String) sets.String {
+	return ready.Difference(total)
+}
```
Comment (RainbowMango) [inline pkg/scheduler/scheduler.go:739]: I can't see the `total` and `ready` meaning if I just look at the function. It's a math thing.

Consider rename it or add some comments.

**Review by RainbowMango — COMMENTED** [review]
I like the two utils, though.

File: pkg/scheduler/scheduler.go (hunk 5 above)
Comment (XiShanYongYe-Chang) [reply pkg/scheduler/scheduler.go]: modify like this:

```go
// bindClusters: set of cluster which exist in resourceBinding/clusterResourceBinding's spec.
// readyClusters: set of cluster whose health status is ready.
// calcReservedCluster calculate set of cluster whose health status is ready in bindClusters.
func calcReservedCluster(bindClusters, readyClusters sets.String) sets.String {
	return bindClusters.Difference(bindClusters.Difference(readyClusters))
}

// bindClusters: set of cluster which exist in resourceBinding/clusterResourceBinding's spec.
// readyClusters: set of cluster whose health status is ready.
// calcAvailableCluster calculate set of cluster whose health status is ready and not in bindClusters.
func calcAvailableCluster(bindCluster, readyClusters sets.String) sets.String {
	return readyClusters.Difference(bindCluster)
}
```

File: pkg/scheduler/scheduler.go [hunk 6]
```diff
@@ -770,13 +723,77 @@ func (s *Scheduler) rescheduleResourceBinding(resourceBinding *workv1alpha1.Reso
...
+func calcReservedCluster(bindClusters, readyClusters sets.String) sets.String {
+	return bindClusters.Difference(bindClusters.Difference(readyClusters))
+}
+
+// bindClusters: set of cluster which exist in resourceBinding/clusterResourceBinding's spec.
+// readyClusters: set of cluster whose health status is ready.
+// calcAvailableCluster calculate set of cluster whose health status is ready and not in bindClusters.
+func calcAvailableCluster(bindCluster, readyClusters sets.String) sets.String {
+	return readyClusters.Difference(bindCluster)
+}
```
Comment (RainbowMango) [inline pkg/scheduler/scheduler.go:745]:
```suggestion
// calcReservedCluster eliminates the not-ready clusters from the 'bindClusters'.
func calcReservedCluster(bindClusters, readyClusters sets.String) sets.String {
	return bindClusters.Difference(bindClusters.Difference(readyClusters))
}

// calcAvailableCluster returns a list of ready clusters that not in 'bindClusters'.
func calcAvailableCluster(bindCluster, readyClusters sets.String) sets.String {
	return readyClusters.Difference(bindCluster)
}
```

**Pull request merged**